ast/desugar/Desugar.cc

```diff
@@ -1059,9 +1059,32 @@ ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what)
                     auto iff = MK::If(loc, std::move(cond), std::move(rhs), std::move(lhs));
                     result = std::move(iff);
                 } else {
-                    core::NameRef tempName = dctx.freshNameUnique(core::Names::andAnd());
-                    auto temp = MK::Assign(loc, tempName, std::move(lhs));
-                    auto iff = MK::If(loc, MK::Local(loc, tempName), std::move(rhs), MK::Local(loc, tempName));
+                    auto andAndTemp = dctx.freshNameUnique(core::Names::andAnd());
+
+                    auto *lhsSend = ast::cast_tree<ast::Send>(lhs);
+                    auto *rhsSend = ast::cast_tree<ast::Send>(rhs);
+
+                    ExpressionPtr thenp;
+                    if (lhsSend != nullptr && rhsSend != nullptr) {
+                        auto lhsSource = core::Loc(dctx.ctx.file, lhsSend->loc).source(dctx.ctx);
+                        auto rhsRecvSource = core::Loc(dctx.ctx.file, rhsSend->recv.loc()).source(dctx.ctx);
+                        if (lhsSource.has_value() && lhsSource == rhsRecvSource) {
```

Comment:
So it looks like we're only inserting `checkAndAnd`s when the lhs and the rhs receiver are lexically identical, right? So for example, `x.foo && x . foo . bar` wouldn't get one. (I think that's probably reasonable, if I'm understanding correctly.)

Reply:
Yeah, that's right. it's not great

Reply:
i added a test for this.